File: cobble-binding/cobble-cpp/cpp/lifecycle.cpp
#include <cobble/database.hpp>

#include "detail/convert.hpp"
#include "detail/error.hpp"
#include "detail/impl.hpp"

namespace cobble {

std::uint32_t Database::NowSeconds() const {
  if (!impl_) {
    throw Error(ErrorCode::kInvalidState, "Database has been moved from");
  }
  return ffi::native_database_now_seconds(*impl_->native);
}

void Database::SwitchMemtableType(MemtableType type,
                                  bool flush_current) const {
  if (!impl_) {
    throw Error(ErrorCode::kInvalidState, "Database has been moved from");
  }
  detail::Translate([&] {
    ffi::native_database_switch_memtable_type(
        *impl_->native, static_cast<std::uint8_t>(type), flush_current);
  });
}

std::size_t Database::LoadReadonlyFilesToPrimary() const {
  if (!impl_) {
    throw Error(ErrorCode::kInvalidState, "Database has been moved from");
  }
  return detail::ToSize(detail::Translate([&] {
                          return ffi::
                              native_database_load_readonly_files_to_primary(
                                  *impl_->native);
                        }),
                        "readonly file count");
}

void Db::SetTime(std::uint32_t unix_seconds) const {
  if (!impl_) {
    throw Error(ErrorCode::kInvalidState, "Db has been moved from");
  }
  ffi::native_sharded_database_set_time(*impl_->native, unix_seconds);
}

std::uint32_t Db::NowSeconds() const {
  if (!impl_) {
    throw Error(ErrorCode::kInvalidState, "Db has been moved from");
  }
  return ffi::native_sharded_database_now_seconds(*impl_->native);
}

void Db::SwitchMemtableType(MemtableType type, bool flush_current) const {
  if (!impl_) {
    throw Error(ErrorCode::kInvalidState, "Db has been moved from");
  }
  detail::Translate([&] {
    ffi::native_sharded_database_switch_memtable_type(
        *impl_->native, static_cast<std::uint8_t>(type), flush_current);
  });
}

std::size_t Db::LoadReadonlyFilesToPrimary() const {
  if (!impl_) {
    throw Error(ErrorCode::kInvalidState, "Db has been moved from");
  }
  return detail::ToSize(detail::Translate([&] {
                          return ffi::
                              native_sharded_database_load_readonly_files_to_primary(
                                  *impl_->native);
                        }),
                        "readonly file count");
}

}  // namespace cobble

File: cobble-binding/cobble-cpp/src/lifecycle.rs
use crate::{
    BridgeResult, database::NativeDatabase, error::format_cobble_error, error::input_error,
};

pub(crate) fn native_database_now_seconds(db: &NativeDatabase) -> u32 {
    db.db.now_seconds()
}

pub(crate) fn native_database_switch_memtable_type(
    db: &NativeDatabase,
    kind: u8,
    flush_current: bool,
) -> BridgeResult<()> {
    let memtable_type = match kind {
        0 => cobble_binding::MemtableType::Hash,
        1 => cobble_binding::MemtableType::Skiplist,
        2 => cobble_binding::MemtableType::Vec,
        3 => cobble_binding::MemtableType::Adaptive,
        _ => return Err(input_error("unknown memtable type")),
    };
    db.db
        .switch_memtable_type(memtable_type, flush_current)
        .map_err(format_cobble_error)
}

pub(crate) fn native_database_load_readonly_files_to_primary(
    db: &NativeDatabase,
) -> BridgeResult<u64> {
    db.db
        .load_readonly_files_to_primary()
        .and_then(|count| {
            u64::try_from(count).map_err(|_| {
                cobble_binding::Error::InvalidState(
                    "readonly file count does not fit in u64".to_string(),
                )
            })
        })
        .map_err(format_cobble_error)
}
